tutorials/math/matrix-exponentiation.md

```diff
@@ -9,8 +9,522 @@ keywords:
+
+$f(len, j) = \text{Number of ways to make a string of length len such that it ends with character j,}$, here $j \in \{a, e, i, o, u\}.$
+
+If we know the current character then we know which characters can come after it. For example, if our string ends at character $\text{'a'}$ then we know that the next character is bound to be $\text{'e'}$. Simiarly, if our string ends at character $\text{'o'}$ then the next character must be either $\text{'i'}$  or $\text{'u'}.$
+
+We try to see the constraints in reverse, i.e., if we want our string of length $n$ to have $j$ as its ending character then what all possible string of length $(n - 1)$ are there which can be extended. Our final answer will be $\sum^{i \in \text{\{a, e, i, o, u\}}} f(n, i)$.
+
+Time Complexity of above solution is $O(n)$, which can be reduced to $O(log(n))$ by using matrix exponentiation. Let us see how we can do so. First let us form the recurrence.
+
+$$$
+f(len, j) = 
+    \begin{cases}
+        f(len - 1,\ e) + f(len,\ i) + f(i,\ u)\ & j = a\\
+        f(len - 1,\ a) + f(len - 1,\ i)\ & j = e\\
+        f(len - 1,\ e) + f(len - 1,\ o)\ & j = i\\
+        f(len - 1,\ i)\ & j = o\\
+        f(len - 1,\ i) + f(len - 1,\ o)\ & j = u
+    \end{cases}
+$$$
+
+One can easily verify the recurrence. For example, if we want our string of length $n$ to end at character $\text{'i'}$ then we can only extend strings which ended at character $\text{'e'}$ or $\text{'o'}$ and which have length $(n - 1)$. 
+
+Now, we have a linear recurrence. For Let us try to find the matrix $T$ such that 
+$$$
+T \cdot
+\begin{bmatrix}
+  f(len - 1, a) \\
+  f(len - 1, e) \\
+  f(len - 1, i) \\
+  f(len - 1, o) \\
+  f(len - 1, u) 
+\end{bmatrix}
+= 
+\begin{bmatrix}
+  f(len, a) \\
+  f(len, e) \\
+  f(len, i) \\
+  f(len, o) \\
+  f(len, u) 
+\end{bmatrix}
+= 
+\begin{bmatrix}
+  \begin{array}{l}
+    f(len - 1, e) + f(len - 1, i) + f(len - 1, u)\\
+    f(len - 1, a) + f(len - 1, i)\\
+    f(len - 1, e) + f(len - 1, o)\\
+    f(len - 1, i) \\
+    f(len - 1, i) + f(len - 1, o)
+  \end{array}
+\end{bmatrix}
+$$$
+
+<details>
+
+<summary> Finding T </summary>
+
+$$$
+T = 
+\begin{bmatrix}
+0 & 1 & 1 & 0 & 1 & \\ 
+1 & 0 & 1 & 0 & 0 & \\ 
+0 & 1 & 0 & 1 & 0 & \\ 
+0 & 0 & 1 & 0 & 0 & \\ 
+0 & 0 & 1 & 1 & 0 & 
+\end{bmatrix}
+$$$
+
+If you are unable to see how we came up with matrix $T$, then for simplicity let us write $f(len - 1, ch)$ as only $ch$. Now $T \cdot A = B$ looks as shown below:
+
+$$$
+T \cdot \begin{bmatrix}
+a \\
+e \\
+i \\
+o \\ 
+u
+\end{bmatrix} 
+= 
+\begin{bmatrix}
+    e + i + u \\
+    a + i \\
+    e + o \\
+    i \\
+    i + o
+\end{bmatrix}
+$$$
+
+</details>
+
+Our `multiply` function has one change that was mentioned at the end of last problem (taking modulo at each operation) and the `power` function remains unchanged.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Ishwarendra"/>
+
+```cpp
+int countVowelPermutation(int n) {
```

Comment:
can you provide the full solution? I got signed integer overflow.

Reply:
> can you provide the full solution? I got signed integer overflow.

Full solution inside the md file or send you separately?

Reply:
[Here](https://leetcode.com/problems/count-vowels-permutation/submissions/1066217338/) is the accepted solution on leetcode. Did you make changes in the `multiply(...)` function?

Reply:
Oh ok. I missed adding MOD in multiply.